### PR TITLE
[Distributed] Add finer granularity tag for distributed submodule

### LINKED ISSUE #113544: P0: Integrate distributed logger with TORCH_LOGS, make sure upper level library can't override distributed LOG level.
Integrate torch distributed logging with TORCH_LOGS, by using a exclusive `distributed` artifact

*Option 1*
'distributed' component covers all of distributed logs (at least python ones)

*Option 2*
use subcomponents (ddp, fsdp, c10d, ...) instead of one major component

*Option 3*
can we do both? I don't know if the infra supports this  (distributed is a parent, ddp, fsdp are children, can be configured separately)

Note: logs from C++ components for distributed also need to be addressed, an i'm not sure if the current TORCH_LOGS infra provides a way to unify those.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @lucasllc @d4l3k

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116434

This PR is the start to enable the integrate pytorch distributed logs in Torch LOGs. We now already have one tag "distributed" for all distributed components but distributed is a very large component and we want to have some hierarchy and give users options to only turn on logs for certain submodules. So we also added tags starting with "dist_*" for each submodule. (This PR only adds some of them and we are going to add more down the road)

Related discussions can be found here: https://github.com/pytorch/pytorch/issues/113544

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225 @fritzo @neerajprad @alicanb @nikitaved

### CODE REVIEW

File: torch/_logging/_internal.py [hunk 1]
```diff
@@ -254,6 +257,18 @@ def set_logs(
             Whether to log communication operations and other debug info from pytorch distributed components.
             Default: ``logging.WARN``
 
+        dist_c10d (:class:`Optional[int]`):
+            Whether to log communication operations related debug info in pytorch distributed components.
```
Comment (awgu) [inline torch/_logging/_internal.py:261]: nit (before this PR): Should we stylize as "PyTorch Distributed"?

**Review by awgu — APPROVED** [review]
This looks reasonable to me!

File: torch/_logging/_internal.py (hunk 1 above)
Comment (wanchaol) [inline torch/_logging/_internal.py:261]: nit: whether to log c10d communication operation

File: torch/_logging/_registrations.py [hunk 2]
```diff
@@ -2,7 +2,11 @@
 from ._internal import register_artifact, register_log
 
 DYNAMIC = ["torch.fx.experimental.symbolic_shapes", "torch.fx.experimental.sym_node"]
-DISTRIBUTED = ["torch.distributed", "torch._dynamo.backends.distributed"]
+DISTRIBUTED = [
+    "torch.distributed",
+    "torch._dynamo.backends.distributed",
+    "torch.nn.parallel.distributed",
```
Comment (wanchaol) [inline torch/_logging/_registrations.py:8]: missing fsdp here?

File: torch/_logging/_registrations.py [hunk 3]
```diff
@@ -11,6 +15,11 @@
 register_log("dynamic", DYNAMIC)
 register_log("torch", "torch")
 register_log("distributed", DISTRIBUTED)
+register_log(
+    "dist_c10d", ["torch.distributed.distributed_c10d", "torch.distributed.rendezvous"]
+)
+register_log("dist_ddp", ["torch.nn.parallel.distributed"])
+register_log("dist_fsdp", ["torch.distributed.fsdp"])
```
Comment (wanchaol) [inline torch/_logging/_registrations.py:22]: I hope we won't add too many top level entries in the future that make TORCH_LOGS hard to use, if that happens we can consider a hierarchical parsing like `distributed.ddp` at that time I guess.

**Review by wanchaol — APPROVED** [review]
lgtm! some nits only

File: torch/_logging/_registrations.py (hunk 2 above)
Comment (fduwjj) [reply torch/_logging/_registrations.py:8]: Oh, fsdp is a child of "torch.distributed", so it should be in already.

File: torch/_logging/_registrations.py (hunk 3 above)
Comment (fduwjj) [reply torch/_logging/_registrations.py:22]: Agree, that's why I didn't add so many here. We should at most have 5 or 6 submodule labels.. The current three should cover most debugging experiences IMHO.

File: torch/_logging/_internal.py (hunk 1 above)
Comment (fduwjj) [reply torch/_logging/_internal.py:261]: Actually, by consulting with ChatGPT, it says:

> Use "operations" if you are referring to multiple actions or procedures, or to the general concept of a series of actions or procedures.

Here we actually want loggings for multiple actions (collectives, rendezvous), so that we keep it same as L257. Let me know if this sounds make sense or not.

Comment (fduwjj) [reply torch/_logging/_internal.py:261]: Agree... Actually, PyTorch part is for sure, I am not sure if we should capitalize distributed. Or we just treat "PyTorch Distributed" as a term?

Comment (fduwjj) [reply torch/_logging/_internal.py:261]: Actually by checking PTD's code base, looks like we already using `PyTorch Distributed`, let's just keep this tradition. (Maybe change all PyTorch distributed to PyTorch Distributed as a follow-up)

Comment (wanchaol) [reply torch/_logging/_internal.py]: oh I meant adding "c10d", not commenting about the "operation" :)

Comment (fduwjj) [reply torch/_logging/_internal.py]: oh, sorry. Missed that, let me send out an another PR for this.
